cli/commands.go

```diff
@@ -283,6 +284,10 @@ func runDaemon(d *app.MenderDaemon) error {
 				d.ForceToState <- app.States.InventoryUpdate
 			} else if s == syscall.SIGTERM {
 				d.StopDaemon()
+				time.AfterFunc(time.Second * 5, func() {
```

Comment:
this at least should be a global variable or const and maybe even consider configuration option?
have we considered sending INT, HUP, STOP, or else to interrupt the operations?
does it mainly concern writing an update?
is it possible that we brick if we os.Exit in the middle of something?
in my book it is a bad practice to do anything else than set an atomic flag in the signal handler.

Reply:
> this at least should be a global variable or const and maybe even consider configuration option?

Const I can agree with, but I think this is too implementation specific to be a configuration option.

> have we considered sending INT, HUP, STOP, or else to interrupt the operations?

The commit message is actually slightly misleading: The primary problem is that this happens while rebooting, and there is no operation going on then, the client is just waiting to be killed from the outside. But since it is waiting inside the reboot state, it is not returning to the state loop where it would exit. Even if we could fix this, we can't fix it for Update Modules, because we don't know what they do inside their scripts.

The hanging *also* happens during Download, as the commit message mentions, but this is not the primary problem.

> does it mainly concern writing an update?
> is it possible that we brick if we os.Exit in the middle of something?

The client is prepared to handle sudden exits (actually, sudden reboots), so exiting in the middle of something is no problem.

> in my book it is a bad practice to do anything else than set an atomic flag in the signal handler.

This isn't a signal handler in the same way as in C, so we don't need to worry about all the atomic stuff.

Reply:
I'm unsure about the timer though. It's a good idea in principle, and for other daemons it would clearly have been the best way. But my concern is that this will happen on absolutely all updates. In fact, we are *guaranteed* to hit this code in a normal update because the client "hangs" inside the reboot state until it is killed from the outside. So essentially the "clean path" will almost never be taken, and I'm wondering if we're just delaying things for no reason. I think my vote would be in favor of no delay, just straight exit, but I'm open to both solutions. What do you think @merlin-northern and @oleorhagen?

Reply:
ok I can understand that, but when I do `systemctl restart mender-client` I would expect it to gracefully handle SIGTERM and exit if possible. I am just not a fan of an idea that we just do exit on SIGTERM regardless of anything. is it possible here to detect somehow that we have to call exit?

Reply:
But which part about calling `os.Exit()` is not graceful? As I said, the client is prepared to handle this scenario (we only use ACID transactions in the database, it reloads state from where it left off, etc).

Reply:
probably only in the part in my head :) just my feeling. if you are ok with os.Exit then I will not argue (too much).

Reply:
In general it's a good feeling to have, and I agree, but this particular daemon is essentially *made for* sudden exits!